lib/web/join_tokens.go

```diff
@@ -239,20 +240,41 @@ func createJoinToken(ctx context.Context, m nodeAPIGetter, roles types.SystemRol
 	}, nil
 }
 
-func getJoinScript(settings scriptSettings, m nodeAPIGetter) (string, error) {
-	// Skip decoding validation for IAM tokens since they are generated with a different method
-	if settings.joinMethod != string(types.JoinMethodIAM) {
-		// This token does not need to be validated against the backend because it's not used to
-		// reveal any sensitive information. However, we still need to perform a simple input
-		// validation check by verifying that the token was auto-generated.
-		// Auto-generated tokens must be encoded and must have an expected length.
-		decodedToken, err := hex.DecodeString(settings.token)
+func getJoinScript(ctx context.Context, settings scriptSettings, m nodeAPIGetter) (string, error) {
+	switch settings.joinMethod {
+	case string(types.JoinMethodUnspecified), string(types.JoinMethodToken), string(types.JoinMethodIAM):
+		if settings.joinMethod != string(types.JoinMethodIAM) {
+			decodedToken, err := hex.DecodeString(settings.token)
+			if err != nil {
+				return "", trace.Wrap(err)
+			}
+			if len(decodedToken) != auth.TokenLenBytes {
+				return "", trace.BadParameter("invalid token %q", decodedToken)
+			}
+		}
+
+		// The provided token can be attacker controlled, so we must validate
+		// it with the backend before using it to generate the script.
+		_, err := m.GetToken(ctx, settings.token)
 		if err != nil {
-			return "", trace.Wrap(err)
+			return "", trace.BadParameter("invalid token")
 		}
+	default:
+		return "", trace.BadParameter("join method %q is not supported via script", settings.joinMethod)
+	}
 
-		if len(decodedToken) != auth.TokenLenBytes {
-			return "", trace.BadParameter("invalid token length")
+	// We must also validate the label spec, which can be controlled by
+	// an attacker and is fed into the join script.
+	if strings.ContainsAny(settings.nodeLabels, "\r\n'#$`\\") {
```

Comment:
Instead of a blocklist, perhaps an allowlist would be better/safer? See https://github.com/alessio/shellescape/blob/master/shellescape.go#L26 for example.

Reply:
We could just stop parsing the header with node labels in it. The UI doesn't (yet) expose this functionality, so if the labels are set, probably someone is doing something unsupported.

We really should move away from this custom format that's easy to inject logic into and serialize labels to a well-known format, which we can do as part of the Teleport Discover work.